File: packages/expo-router/android/src/main/java/expo/modules/router/ModalPortalContentWrapperView.kt
package expo.modules.router

import android.content.Context
import android.util.Log
import com.facebook.react.views.view.ReactViewGroup
import java.lang.ref.WeakReference

class ModalPortalContentWrapperView(context: Context) :
  ReactViewGroup(context) {
  private var host: WeakReference<ModalPortalHostView?> = WeakReference(null)
  private var contentView: WeakReference<ModalPortalContentView?> = WeakReference(null)

  fun setContentView(child: ModalPortalContentView){
    if (contentView.get() != null) {
      Log.w(
        "ExpoRouter",
        "Warning: Multiple ModalPortalContentView components found. Only the first one will be used."
      )
    }
    host.get()?.setContentView(child)
    contentView = WeakReference(child)
  }

  fun unsetContentView() {
    val contentView = this.contentView.get()
    if (contentView == null) {
      Log.w(
        "ExpoRouter",
        "Warning: Removing child that does not exist from ModalPortalContentWrapperView"
      )
    } else {
      this.host.get()?.unmountContentView()
    }
    this.contentView = WeakReference(null)
  }

  fun setHostId(hostId: String) {
    val hostView = PortalHostsRegistry.getHost(hostId)
    if (hostView != null) {
      host = WeakReference(hostView)
      val contentView = this.contentView.get()
      if (contentView != null) {
        hostView.setContentView(contentView)
      }
    } else {
      Log.w("ExpoRouter", "Host view with id $hostId not found")
    }
  }
}

File: packages/expo-router/ios/Portal/ModalPortalContentWrapperView.swift
import ExpoModulesCore
import React

class ModalPortalContentWrapperView: ExpoView {
  var host: ModalPortalHostView?
  var contentView: ModalPortalContentView?

  required init(appContext: AppContext? = nil) {
    super.init(appContext: appContext)
    clipsToBounds = true
  }

  override func mountChildComponentView(_ childComponentView: UIView, index: Int) {
    if contentView != nil {
      print(
        "Warning: Multiple ModalPortalContentView components found. Only the first one will be used."
      )
      return
    }
    if let content = childComponentView as? ModalPortalContentView {
      self.contentView = content
      print("Mounting content view")
      self.host?.unmountContentView()
      self.host?.setContentView(contentView: content)
    } else {
      print("Mounting: Child component view must be of type ModalPortalContentView")
    }
  }

  override func unmountChildComponentView(_ childComponentView: UIView, index: Int) {
    if let content = childComponentView as? ModalPortalContentView {
      print("Unmounting content view")
      self.host?.unmountContentView()
      contentView = nil
    } else {
      print("Unmounting: Child component view must be of type ModalPortalContentView")
    }
  }

  func setHost(hostId: String) {
    let hostView = PortalHostsRegistry.shared.getHost(hostId: hostId)
    if hostView != nil {
      self.host = hostView
      if let contentView = self.contentView {
        self.host?.setContentView(contentView: contentView)
      }
    } else {
      print("Host view not found")
    }
  }
}
